util/winutil/gp/policylock_windows.go

```diff
@@ -0,0 +1,275 @@
+			// Decrement the counter if the lock cannot be acquired,
+			// and complete the pending close request if we're the last owner.
+			if l.lockCnt.Add(-2) == 0 {
+				l.closeInternal()
+			}
+		}
+	}()
+
+	if l.handle.Load() != 0 {
+		// The underlying CriticalPolicySection is already acquired.
+		// It is an R-Lock (with the W-counterpart owned by the Group Policy service),
+		// meaning that it can be acquired by multiple readers simultaneously.
+		// So we can just return.
+		return nil
+	}
+
+	if l.lockCnt.Load()&1 == 0 {
+		// The Close method was called while we waited on l.mu.
+		return ErrInvalidLockState
+	}
+
+	// In some cases in production environments, the Group Policy service may
+	// hold the corresponding W-Lock for extended periods of time (minutes
+	// rather than seconds or milliseconds). We need to make our wait operation
+	// cancellable. So, if one goroutine invokes (*PolicyLock).Close while another
+	// initiates (*PolicyLock).Lock and waits for the underlying R-lock to be
+	// acquired by enterCriticalPolicySection, the Close method should cancel
+	// the wait.
+
+	initCh := make(chan error)
+	resultCh := make(chan policyLockResult)
+
+	go func() {
+		if l.scope == UserPolicy && l.token != 0 {
+			// Impersonate the user whose critical policy section we want to acquire.
+			runtime.LockOSThread()
+			defer runtime.UnlockOSThread()
+			if err := impersonateLoggedOnUser(l.token); err != nil {
+				initCh <- err
+				return
+			}
+			defer func() {
+				if err := windows.RevertToSelf(); err != nil {
+					// RevertToSelf errors are non-recoverable.
+					panic(fmt.Errorf("could not revert impersonation: %w", err))
+				}
+			}()
+		}
+		close(initCh)
+
+		var machine bool
+		if l.scope == MachinePolicy {
+			machine = true
+		}
+		handle, err := enterCriticalPolicySection(machine)
+
+		select {
+		case resultCh <- policyLockResult{handle, err}:
+			// lockSlow has received the result.
+		default:
+			// The lock is being closed, and we lost the race to l.closing.
+			if err == nil {
+				leaveCriticalPolicySection(handle)
+			}
+		}
+	}()
+
+	// lockSlow should not return until the goroutine above has been fully initialized,
+	// even if the lock is being closed.
+	if err = <-initCh; err != nil {
+		return err
+	}
+
+	select {
+	case result := <-resultCh:
+		if result.err == nil {
+			l.handle.Store(uintptr(result.handle))
+		}
+		return result.err
+	case <-l.closing:
+		return ErrInvalidLockState
+	}
+}
+
+// Unlock unlocks l.
+// It panics if l is not locked on entry to Unlock.
+func (l *PolicyLock) Unlock() {
+	lockCnt := l.lockCnt.Add(-2)
+	if lockCnt < 0 {
+		panic("negative lockCnt")
+	}
+	if lockCnt > 1 {
+		// The lock is still being used by other readers.
+		// We compare against 1 rather than 0 because the least significant bit
+		// of lockCnt indicates that l has been initialized and a close
+		// has not been requested yet.
+		return
+	}
+
+	if handle := policyLockHandle(l.handle.Swap(0)); handle != 0 {
```

Comment:
I think there's a race here:

1. Goroutine A calls Lock() and it returns; l.lockCnt == 3
2. Goroutine A calls Unlock() and the decrement occurs, l.lockCnt == 1
3. Goroutine B calls Lock(). l.lockCnt is incremented back to 3, and it returns quickly because l.handle.Load() != 0.
4. Goroutine A continues Unlock() and swaps l.handle with 0. It leaves the critical section.

Now Goroutine B thinks it's in a critical section but it is not.

Reply:
Good catch, thank you.